sql/core/src/main/scala/org/apache/spark/sql/execution/subquery.scala

```diff
@@ -55,6 +55,112 @@ object ExecSubqueryExpression {
   }
 }
 
+/**
+ * Exists is used to test for the existence of any record in a subquery.
+ *
+ * This is the physical copy of Exists to be used inside SparkPlan.
+ */
+case class Exists(
+    plan: BaseSubqueryExec,
+    exprId: ExprId)
+  extends ExecSubqueryExpression {
+
+  override def dataType: DataType = BooleanType
+  override def children: Seq[Expression] = Nil
+  override def nullable: Boolean = false
+  override def toString: String = plan.simpleString(SQLConf.get.maxToStringFields)
+  override def withNewPlan(plan: BaseSubqueryExec): Exists = copy(plan = plan)
+
+  // Whether the subquery returns one or more records
+  @volatile private var result: Boolean = _
+  @volatile private var updated: Boolean = false
+
+  def updateResult(): Unit = {
+    val rows = plan.executeCollect()
+    result = rows.nonEmpty
+    updated = true
+  }
+
+  override def eval(input: InternalRow): Boolean = {
+    require(updated, s"$this has not finished")
+    result
+  }
+
+  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    require(updated, s"$this has not finished")
+    Literal.create(result, BooleanType).doGenCode(ctx, ev)
+  }
+}
+
+/**
+ * Evaluates to `true` if `values` are returned in the subquery's result set.
+ * If `values` are not found in the subquery's result set, and there are nulls in
+ * `values` or the result set, it should return null.
+ * This is the physical copy of InSubquery to be used inside SparkPlan.
+ */
+case class InSubquery(
+    values: Seq[Literal],
+    plan: BaseSubqueryExec,
+    exprId: ExprId)
+  extends ExecSubqueryExpression {
+  override def dataType: DataType = BooleanType
+  override def children: Seq[Expression] = Nil
+  override def nullable: Boolean = true
+  override def toString: String = plan.simpleString(SQLConf.get.maxToStringFields)
+  override def withNewPlan(plan: BaseSubqueryExec): InSubquery = copy(plan = plan)
+
+  @volatile private var result: Boolean = _
+  @volatile private var isNull: Boolean = false
+  @volatile private var updated: Boolean = false
+
+  def updateResult(): Unit = {
+    val rows = plan.executeCollect()
+    // The semantic of '(a,b) in ((x1, y1), (x2, y2), ...)' is
+    // '(a = x1 and b = y1) or (a = x2 and b = y2) or ...'
+    val expression = rows.map(row => {
```

Comment:
@francis0407 I am really sorry... as i raised this earlier. I am just very afraid of this exploding on a large result set. Are we able to fix this in a follow-up in 3.0 ? If not, we should perhaps keep this under a config ? What do you think @cloud-fan ?

Reply:
When we use a `InSubquery`, it is usually executed using a `Broadcast Semi/Anti Join`. I think in that case, the result set is also collected in the driver for broadcasting?

Reply:
@francis0407 It should only use broadcast if the data size is within broadcast threshold to the best of my knowledge.

Reply:
Yeah, I got it. So we still need a optimize rule, I think. We can push the IN condition into the subquery, and use a distinct after the subquery:
```sql
select *
from t
where 3 in (select a from b)
```

The query above can be rewritten as:
```sql
select *
from t
where 3 in (select distinct a 
            from b 
            where a = 3 or a is null)
```
How about this? @dilipbiswal

Reply:
agree with @dilipbiswal that it's a bad idea to create a complex expression tree which is proportional to the size of data.

One idea is to create a special expression for this equal check, instead of creating a lot of `EqualTo` expressions.

Reply:
Yeah, we should use a special expression. I will update this soon.

Reply:
I have updated this, could you please help check it? cc @dilipbiswal  @cloud-fan